File: backend/services/habit_voice.py
"""Coaching copy for habit logging surfaces.

Pure function — no database access, no mutations, no global state.
All coaching strings for the quick-log, streak-feedback, and milestone
surfaces are produced here so no surface hard-codes its own copy.

Public API
----------
compose_log_feedback(habit, week_done, week_target, total_logs, is_miss,
                     current_streak=0)
    Returns a dict with keys:
        message        (str)  — rendered coaching copy
        framing        (str)  — one of: "routine", "weekly", "streak",
                                "milestone", "miss"
        show_identity  (bool) — True only at milestone/landmark thresholds
        next_milestone (int|None) — the next landmark count, or None
"""

from __future__ import annotations

# Milestone log-count landmarks (ascending order)
_MILESTONES = [10, 25, 30, 50, 75, 100, 150, 200, 250, 365, 500, 750, 1000]

# Streak landmarks that unlock identity framing
_STREAK_LANDMARKS = {7, 14, 21, 30, 60, 90, 180, 365}


def _next_milestone(total: int) -> int | None:
    """Return the next milestone count above total, or None if beyond the list."""
    for m in _MILESTONES:
        if m > total:
            return m
    return None


def _is_milestone(total: int) -> bool:
    return total in set(_MILESTONES)


def _is_weekly_habit(habit) -> bool:
    """True when the habit's cadence is less than 7 times per week."""
    # Primary signal: schedule_type == 'times_per_week'
    if getattr(habit, "schedule_type", None) == "times_per_week":
        return True
    # Fallback: weekly_target explicitly set below 7
    wt = getattr(habit, "weekly_target", None)
    if wt is not None and float(wt) < 7:
        return True
    return False


def compose_log_feedback(
    habit,
    week_done: int | None,
    week_target: int | float | None,
    total_logs: int,
    is_miss: bool,
    current_streak: int = 0,
) -> dict:
    """Return coaching copy for one habit log event.

    Parameters
    ----------
    habit:
        Habit row (or SimpleNamespace) with ``name``, ``weekly_target``,
        ``tracking_type``, ``schedule_type``, ``schedule_target``.
    week_done:
        How many times the habit was completed in the current week
        (sourced from the API, never inferred client-side).
    week_target:
        The habit's weekly target (from ``habit.weekly_target``).
    total_logs:
        Total all-time log count for this habit.
    is_miss:
        True when the feedback event is triggered by a recognised missed day.
    current_streak:
        The current consecutive-periods streak (0 by default for non-streak
        events).

    Returns
    -------
    dict
        message        (str)
        framing        (str)  routine | weekly | streak | milestone | miss
        show_identity  (bool)
        next_milestone (int | None)
    """
    habit_name = getattr(habit, "name", "this habit")

    # ── Milestone detection ──────────────────────────────────────────────────
    if _is_milestone(total_logs):
        nxt = _next_milestone(total_logs)
        msg = f"You've logged {habit_name} {total_logs} times."
        if nxt is not None:
            msg += f" Next landmark: {nxt}."
        return {
            "message": msg,
            "framing": "milestone",
            "show_identity": True,
            "next_milestone": nxt,
        }

    # ── Miss acknowledgement ─────────────────────────────────────────────────
    if is_miss:
        wd = int(week_done) if week_done is not None else 0
        wt = int(week_target) if week_target is not None else 7
        msg = f"Yesterday slipped—{wd} of {wt} this week still puts you ahead of most."
        return {
            "message": msg,
            "framing": "miss",
            "show_identity": False,
            "next_milestone": None,
        }

    # ── Streak landmark check ────────────────────────────────────────────────
    if current_streak in _STREAK_LANDMARKS and current_streak > 0:
        wd = int(week_done) if week_done is not None else 0
        wt = int(week_target) if week_target is not None else 7
        msg = (
            f"{current_streak}-period streak on {habit_name}. "
            f"You are becoming someone who does this consistently."
        )
        return {
            "message": msg,
            "framing": "streak",
            "show_identity": True,
            "next_milestone": None,
        }

    # ── Weekly-cadence framing (< 7×/week habits) ────────────────────────────
    if _is_weekly_habit(habit):
        wd = int(week_done) if week_done is not None else 0
        wt_raw = week_target if week_target is not None else getattr(habit, "weekly_target", None)
        wt = int(wt_raw) if wt_raw is not None else 7
        if wd >= wt:
            msg = f"On track—{wd} of {wt} this week."
        else:
            msg = f"{wd} of {wt} this week. Keep going."
        return {
            "message": msg,
            "framing": "weekly",
            "show_identity": False,
            "next_milestone": None,
        }

    # ── Routine daily log ────────────────────────────────────────────────────
    return {
        "message": "Logged. Keep it going.",
        "framing": "routine",
        "show_identity": False,
        "next_milestone": None,
    }

File: frontend/js/lib/habit-voice.js
/**
 * habit-voice.js — Coaching copy for habit logging surfaces (issue #920).
 *
 * All coaching strings for the quick-log, streak-feedback, and milestone
 * surfaces are produced here. No surface hard-codes its own copy strings.
 *
 * Mirror of backend/services/habit_voice.py — same branching logic so
 * frontend rendering stays in sync with backend-computed framing.
 *
 * Public API
 * ----------
 * HabitVoice.compose(habit, weekDone, weekTarget, totalLogs, isMiss, currentStreak)
 *   → { message, framing, showIdentity, nextMilestone }
 *
 * HabitVoice.identityLine(habitName)
 *   → string for "you are becoming" copy
 */
(function (root) {
  'use strict';

  var _MILESTONES = [10, 25, 30, 50, 75, 100, 150, 200, 250, 365, 500, 750, 1000];
  var _MILESTONE_SET = new Set(_MILESTONES);
  var _STREAK_LANDMARKS = new Set([7, 14, 21, 30, 60, 90, 180, 365]);

  function _nextMilestone(total) {
    for (var i = 0; i < _MILESTONES.length; i++) {
      if (_MILESTONES[i] > total) return _MILESTONES[i];
    }
    return null;
  }

  function _isWeeklyHabit(habit) {
    if (habit.schedule_type === 'times_per_week') return true;
    var wt = habit.weekly_target != null ? parseFloat(habit.weekly_target) : null;
    return wt !== null && wt < 7;
  }

  /**
   * Compose coaching copy for one habit log event.
   *
   * @param {object} habit        - Habit row from /api/habits/summary
   * @param {number} weekDone     - Completions in current Mon–Sun week (from API)
   * @param {number|null} weekTarget  - Habit's weekly target (from API)
   * @param {number} totalLogs    - All-time log count (from API)
   * @param {boolean} isMiss      - True when triggered by a recognised missed day
   * @param {number} currentStreak - Current streak length (default 0)
   * @returns {{ message: string, framing: string, showIdentity: boolean, nextMilestone: number|null }}
   */
  function compose(habit, weekDone, weekTarget, totalLogs, isMiss, currentStreak) {
    var name = (habit && habit.name) ? habit.name : 'this habit';
    var streak = currentStreak || 0;
    var wd = weekDone != null ? Math.floor(weekDone) : 0;
    var wt = weekTarget != null ? Math.floor(weekTarget) : 7;

    // ── Milestone detection ──────────────────────────────────────────────────
    if (_MILESTONE_SET.has(totalLogs)) {
      var nxt = _nextMilestone(totalLogs);
      var msg = 'You’ve logged ' + name + ' ' + totalLogs + ' times.';
      if (nxt !== null) msg += ' Next landmark: ' + nxt + '.';
      return { message: msg, framing: 'milestone', showIdentity: true, nextMilestone: nxt };
    }

    // ── Miss acknowledgement ─────────────────────────────────────────────────
    if (isMiss) {
      var missMsg = 'Yesterday slipped—' + wd + ' of ' + wt +
        ' this week still puts you ahead of most.';
      return { message: missMsg, framing: 'miss', showIdentity: false, nextMilestone: null };
    }

    // ── Streak landmark ──────────────────────────────────────────────────────
    if (streak > 0 && _STREAK_LANDMARKS.has(streak)) {
      var sMsg = streak + '-period streak on ' + name + '. ' +
        'You are becoming someone who does this consistently.';
      return { message: sMsg, framing: 'streak', showIdentity: true, nextMilestone: null };
    }

    // ── Weekly-cadence framing (< 7×/week) ───────────────────────────────────
    if (_isWeeklyHabit(habit)) {
      var wMsg = wd >= wt
        ? 'On track—' + wd + ' of ' + wt + ' this week.'
        : wd + ' of ' + wt + ' this week. Keep going.';
      return { message: wMsg, framing: 'weekly', showIdentity: false, nextMilestone: null };
    }

    // ── Routine daily log ────────────────────────────────────────────────────
    return { message: 'Logged. Keep it going.', framing: 'routine', showIdentity: false, nextMilestone: null };
  }

  /**
   * Return identity-framing copy for a habit name.
   * Only call this when showIdentity is true.
   */
  function identityLine(habitName) {
    var n = habitName || 'this habit';
    return 'You are becoming someone who prioritises ' + n + '.';
  }

  root.HabitVoice = { compose: compose, identityLine: identityLine };
}(window));
